ted_sws/data_sampler/services/notice_xml_indexer.py

```diff
@@ -0,0 +1,171 @@
+    unique_notice_ids = notice_ids.copy()
+    notice_repository = NoticeRepository(mongodb_client=mongodb_client)
+    while len(unique_notice_ids):
+        tmp_result = list(notice_repository.collection.aggregate([
+            {"$unwind": "$xml_metadata.unique_xpaths"},
+            {"$match": {
+                "xml_metadata.unique_xpaths": {"$nin": minimal_set_of_xpaths},
+                "ted_id": {"$in": unique_notice_ids}
+            }
+            },
+            {"$group": {"_id": "$xml_metadata.unique_xpaths", "count": {"$sum": 1},
+                        "notice_ids": {"$push": "$ted_id"}}},
+            {"$sort": {"count": -1}},
+            {"$limit": 1}
+        ]))[0]
+        minimal_set_of_xpaths.append(tmp_result["_id"])
+        for notice_id in tmp_result["notice_ids"]:
+            unique_notice_ids.remove(notice_id)
+
+    return minimal_set_of_xpaths
+
+
+def get_minimal_set_of_notices_for_coverage_xpaths(xpaths: List[str], mongodb_client: MongoClient) -> List[str]:
+    minimal_set_of_notices = []
+    unique_xpaths = xpaths.copy()
+    notice_repository = NoticeRepository(mongodb_client=mongodb_client)
+    while len(unique_xpaths):
+        tmp_result = list(notice_repository.collection.aggregate([
+            {"$match": {
+                "ted_id": {"$nin": minimal_set_of_notices},
+            }
+            },
+            {"$unwind": "$xml_metadata.unique_xpaths"},
+            {"$match": {
+                "xml_metadata.unique_xpaths": {"$in": unique_xpaths},
+            }
+            },
+            {"$group": {"_id": "$ted_id", "count": {"$sum": 1}, "xpaths": {"$push": "$xml_metadata.unique_xpaths"}}},
+            {"$sort": {"count": -1}},
+            {"$limit": 1}
+        ]))[0]
+        minimal_set_of_notices.append(tmp_result["_id"])
+        for xpath in tmp_result["xpaths"]:
+            unique_xpaths.remove(xpath)
+
+    return minimal_set_of_notices
+
+
+def get_unique_notices_id_covered_by_xpaths(xpaths: List[str], mongodb_client: MongoClient) -> List[str]:
+    notice_repository = NoticeRepository(mongodb_client=mongodb_client)
+    results = list(notice_repository.collection.aggregate([
+        {"$match": {"xml_metadata.unique_xpaths": {"$in": xpaths}}},
+        {
+            "$group": {"_id": None,
+                       "ted_ids": {"$push": "$ted_id"}
+                       }
+        },
+        {
+            "$project": {
+                "_id": 0,
+                "ted_ids": {
+                    "$setUnion": {
+                        "$reduce": {
+                            "input": '$ted_ids',
+                            "initialValue": [],
+                            "in": {"$concatArrays": ['$$value', '$$this']}
+                        }
+                    }
+                }
+            }
+        }
+    ]))[0]["ted_ids"]
+    return results
+
+
+def get_unique_xpaths_covered_by_notices(notice_ids: List[str], mongodb_client: MongoClient) -> List[str]:
```

Comment:
I hope we have good tests for this.

We also need documentation for each of the functions above.